### PR TITLE
gguf-py : fix some metadata name extraction edge cases

### PR DESCRIPTION
**Pull request by compilade**
Should fix the problem mentioned in <https://github.com/ggerganov/llama.cpp/pull/8579#issuecomment-2238673198> by @maziyarpanahi, which was caused by #7499 not pruning the empty parts of a name.

I've also made `convert_lora_to_gguf.py` use the LoRA adapter directory for the model card path.

# TODO

- [x] Test conversion with a LoRA adapter with `convert_lora_to_gguf.py` and check if the metadata is correct.

---

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low

### CODE REVIEW

**Review by mofosyne — APPROVED** [review]

File: gguf-py/gguf/metadata.py
Comment (mofosyne) [inline gguf-py/gguf/metadata.py:245]: Was not too easy to understand. Got ChatGPT to rewrite it to make sense of it. Seems to checkout but would prefer this to be more grokkable

```python
# Check if there's any size label that includes a number
contains_number_based_size = any(
    any(char.isdecimal() for char in part) 
    for part, types in zip(name_parts, name_types) 
    if "size_label" in types
)

# If a number-based size label is found, process the size labels
if contains_number_based_size:
    for part, types in zip(name_parts, name_types):
        # Only consider entries that have a size label
        if "size_label" in types:
            # Check if the size label is purely alphabetical
            is_alpha_size_label = all(char.isalpha() for char in part)
            # Remove the size label if it's purely alphabetical
            if is_alpha_size_label:
                types.remove("size_label")
```

Comment (compilade) [reply gguf-py/gguf/metadata.py:245]: I don't think putting conditions in variables, then checking if that variable is `True` makes anything clearer. It's an unnecessary indirection.

This part especially:

```python3
            # Check if the size label is purely alphabetical
            is_alpha_size_label = all(char.isalpha() for char in part)
            # Remove the size label if it's purely alphabetical
            if is_alpha_size_label:
                types.remove("size_label")
```
And the comments are extremely redundant with the code.

But I can understand why

```python3
any(c.isdecimal() for n, t in zip(name_parts, name_types) if "size_label" in t for c in n)
```
might not be obvious at first.

I first got introduced to that syntax with

https://github.com/ggerganov/llama.cpp/blob/87e397d00bdcedd5cbf6dfda06a7b0f302462728/convert_hf_to_gguf.py#L1559-L1560

The way to see it is that the second `for` is inside the first, while the expression before all the `for`s is as if it was in the innermost `for`.

It's explained in <https://docs.python.org/3.12/reference/expressions.html#generator-expressions> and <https://docs.python.org/3.12/reference/expressions.html#displays-for-lists-sets-and-dictionaries>.

**Comment by compilade** [comment]
@mofosyne Regarding the use of title case for the `finetune` part when generating the "standard" name,

https://github.com/ggerganov/llama.cpp/blob/87e397d00bdcedd5cbf6dfda06a7b0f302462728/gguf-py/gguf/utility.py#L61

I think it looks a bit weird for a few models I've recently tried converting:

- `Gemma-2-9B-It-F16.gguf` (seems from Italy instead of instruct-tuned)
  - with original case for `finetune`: `Gemma-2-9B-it-F16.gguf`
  - with original case for the `basename` too: `gemma-2-9B-it-F16.gguf`
- `Gemma-2B-Zephyr-Dpo-F16.gguf` (`dpo` should not be capitalized)
  - with original case for `finetune`: `Gemma-2B-zephyr-dpo-F16.gguf`
  - with orignial case for the `basename` too: `gemma-2B-zephyr-dpo-F16.gguf`

Should the original case be preserved for `finetune`?
What about the `basename` part?

Do you think I should change this in this PR or another one?

---

(out of scope, but still want to comment about this)

Also note that the naming convention at <https://github.com/ggerganov/ggml/blob/master/docs/gguf.md#gguf-naming-convention> should be updated to reflect the format used in #7499 (especially the `finetune` part which is missing in the spec). If the word-only size labels are too complicated for the spec, they could still be detected when converting, but replaced with the calculated size label instead. 

Also I don't think we're close to having any model with Quadrillions of parameter so I suggest removing this, especially since there is no example of what suffix is normally used for that case. (might be `P` for `peta-` as in the SI prefixes)

Let me know what you think about this.

**Comment by mofosyne** [comment]
@compilade well if it's an edge case then it's part of the PR, depending on how urget this fix needs to get in. Is it breaking anyone's workflow currently, if not then we can take our time making it better.

I don't think there is any one right answer here... but we can assume anything read to the kv store is of the correct capitalisation... as for anything heuristically obtained... it's the wild west.

**Comment by mofosyne** [comment]
@compilade and yes the gguf doc is outdated and from what I understand the consensus is that code is the main spec, so our implementation supersedes the document. We should update it.

Also I don't think the heuristics approach we have is a spec, it's a nice to have feature in the converter python. But the gguf spec for naming should contain only what we are aiming for people to stick to.

As for the Quadrillion, I'll suggest keeping it in... I would generally like to have +1 ahead of whatever we can immediately conceptionalise in the near future. And the near future is likely Trillion so +1 would be quadrillion.

**Comment by compilade** [comment]
> And the near future is likely Trillion so +1 would be quadrillion.

I think a quadrillion can safely be 1000T instead.

Personally I think we're quite far from making models with 10 times more parameters as the number of neurons in a human brain.

> Also I don't think the heuristics approach we have is a spec, it's a nice to have feature in the converter python. But the gguf spec for naming should contain only what we are aiming for people to stick to.

Totally agree. The heuristics are there to turn non-compliant but commonly used names into compliant names. The spec only needs to deal with the resulting names, not the mess they come from.

**Comment by mofosyne** [comment]
> @mofosyne
> 
> I think a quadrillion can safely be 1000T instead.
> 
> Personally I think we're quite far from making models with 10 times more parameters than the number of neurons in a human brain.

Well people in the past thought that 640K is ought to be enough for anyone and I like to be optimistic for the development of humanity that we can create something like that in the future. Not going to fight to hard to keep it, but I'll rather it stay.

> > Also I don't think the heuristics approach we have is a spec, it's a nice to have feature in the converter python. But the gguf spec for naming should contain only what we are aiming for people to stick to.
> 
> Totally agree. The heuristics are there to turn non-compliant but commonly used names into compliant names. The spec only needs to deal with the resulting names, not the mess they come from.

Glad we are on the same page... just gotta remember to do it

**Comment by mofosyne** [comment]
Oh noticed there is this todo

> Test conversion with a LoRA adapter with convert_lora_to_gguf.py and check if the metadata is correct.

Not familiar with this so will hold off and let @compilade handle it

**Review by mofosyne — APPROVED** [review]

**Comment by compilade** [comment]
I've tested LoRA generation, and it seems fine.

Here's some `gguf-dump.py` output for <https://huggingface.co/grimjim/Llama-3-Instruct-abliteration-LoRA-8B>:
```
INFO:gguf-dump:* Loading: /srv/LLMstash/tmp/metadata-names/Llama-3-8B-Instruct-abliteration-F16-LoRA.gguf
* File is LITTLE endian, script is running on a LITTLE endian host.
* Dumping 42 key/value pair(s)
      1: UINT32     |        1 | GGUF.version = 3
      2: UINT64     |        1 | GGUF.tensor_count = 450
      3: UINT64     |        1 | GGUF.kv_count = 39
      4: STRING     |        1 | general.architecture = 'llama'
      5: STRING     |        1 | general.type = 'adapter'
      6: STRING     |        1 | adapter.type = 'lora'
      7: STRING     |        1 | general.name = 'Llama 3 Instruct Abliteration LoRA 8B'
      8: STRING     |        1 | general.finetune = 'Instruct-abliteration'
      9: STRING     |        1 | general.basename = 'Llama-3'
     10: STRING     |        1 | general.size_label = '8B'
     11: STRING     |        1 | general.license = 'llama3'
     12: UINT32     |        1 | general.base_model.count = 2
     13: STRING     |        1 | general.base_model.0.name = 'Meta Llama 3 8B Instruct'
     14: STRING     |        1 | general.base_model.0.organization = 'Meta Llama'
     15: STRING     |        1 | general.base_model.0.repo_url = 'https://huggingface.co/meta-llama/Meta-Llama-3-8B-Instruct'
     16: STRING     |        1 | general.base_model.1.name = 'Meta Llama 3 8B Instruct Abliterated v3'
     17: STRING     |        1 | general.base_model.1.version = 'v3'
     18: STRING     |        1 | general.base_model.1.organization = 'Failspy'
     19: STRING     |        1 | general.base_model.1.repo_url = 'https://huggingface.co/failspy/Meta-Llama-3-8B-Instruct-abli'
     20: [STRING]   |        2 | general.tags
     21: FLOAT32    |        1 | adapter.lora.alpha = 32.0
...(then follows hparams and the tokenizer)...
```

There's still some title-casing of the metadata which I didn't remove, but at least it's easy to see that the metadata from the LoRA adapter's model card was correctly used to link to the base models (even if the complete links are truncated in the dump output).

I don't think the hparams and the tokenizer are *really* used in GGUF LoRA adapters though, but that's out of the scope of this PR.

I consider this merge ready, and will merge after #8597.

**Comment by compilade** [comment]
@mofosyne, I recently tried converting <https://huggingface.co/HuggingFaceTB/SmolLM-135M-Instruct>, and got surprised that the default name chosen for `--outtype q8_0` is `cosmo2-135M-webinst-sc2-Q8_0.gguf`.

I think maybe `_name_or_path` from `config.json` should not be used for the name extraction. The whole SmolLM suite seems to use weird `_name_or_path` values, especially the non-Instruct variants.

<https://huggingface.co/HuggingFaceTB/SmolLM-135M-Instruct>:
```json
  "_name_or_path": "HuggingFaceTB/cosmo2-135M-webinst-sc2",
```

<https://huggingface.co/HuggingFaceTB/SmolLM-135M>:
```json
  "_name_or_path": "/fsx/elie_bakouch/checkpoints/final-149M/600000",
```

**Comment by mofosyne** [comment]
@compilade urgh, perhaps we should remove _name_or_path from the heuristics

**Comment by compilade** [comment]
> But this may be an argument for us 'defining' the model card standard in terms of parsing from it... as the huggingface current approach is pretty barebone. This would then encourage model card writers to start filling in their model card with more metadata perhaps?

@mofosyne

Yes, using the model card metadata for overrides seems better than a `metadata.json` tied to the internal GGUF metadata key names, especially if the metadata is useful for more than only `llama.cpp`-related projects.

But there are 2 model cards to consider: the original model card, and the model card of the converted model, which I guess could be taken from the directory where the GGUF model is exported,  but some people use a single directory for all their conversions, which might be problematic for fields like the `license`.

Something else which could be useful is to offer an option to *generate* an initial model card for GGUF conversions based on the original model card with some extra metadata deduced from the model and/or update the metadata of an existing model card.
There are downsides with generating a model card, however, like users not always modifying it afterwards.

**Pull request merged**
